File: 9.贪心算法/12.分发糖果/135.分发糖果.py
#
# @lc app=leetcode.cn id=135 lang=python3
#
# [135] 分发糖果
#
from mytools import *
# @lc code=start
class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        if n == 1: return 1
        res = [1] * n
        for i in range(1, n):
            if ratings[i] > ratings[i - 1]:
                res[i] = res[i - 1] + 1
        for i in range(n - 2, -1, -1):
            if ratings[i] > ratings[i + 1]:
                res[i] = max(res[i + 1] + 1, res[i])
        return sum(res)
# @lc code=end

File: 9.贪心算法/12.分发糖果/135.分发糖果.cpp
/*
 * @lc app=leetcode.cn id=135 lang=cpp
 *
 * [135] 分发糖果
 */
#include "tools.h"
// @lc code=start
class Solution
{
public:
    int candy(vector<int> &ratings)
    {
        int n = ratings.size();

        vector<int> l(n, 1);
        vector<int> r(n, 1);
        for (int i = 1; i < n; i++)
        {
            if (ratings[i] > ratings[i - 1])
                l[i] = l[i - 1] + 1;
        }
        for (int i = n - 2; i >= 0; i--)
        {
            if (ratings[i] > ratings[i + 1])
                r[i] = r[i + 1] + 1;
        }
        int sums = 0;
        for (int i = 0; i < n; i ++)
        {
            sums += max(l[i], r[i]);
        }
        return sums;
    }
};
// @lc code=end
